examples_tests/51.RadixSort/main.cpp

```diff
@@ -0,0 +1,269 @@
+	auto device = createDeviceEx(params);
+
+	if (!device)
+		return 1;
+
+	IVideoDriver* driver = device->getVideoDriver();
+
+	io::IFileSystem* filesystem = device->getFileSystem();
+	asset::IAssetManager* am = device->getAssetManager();
+
+	unsigned int seed = 128;
+
+	// Create (an almost) 256MB input buffer
+	const size_t in_count = (1 << 25) - 23;
+	const size_t in_size = in_count * sizeof(SortElement);
+
+	std::cout << "Input element count: " << in_count << std::endl;
+
+	SortElement* in = new SortElement[in_count];
+	srand(seed++);
+	for (size_t i = 0u; i < in_count; ++i)
+	{
+		in[i].key = rand();
+		in[i].data = i;
+	}
+	
+	auto in_gpu = driver->createFilledDeviceLocalGPUBufferOnDedMem(in_size, in);
+
+	// Take (an almost) 64MB portion from it to sort
+	size_t begin = (1 << 23) + 112;
+	size_t end = (1 << 24) - 77;
+
+	SBufferRange<IGPUBuffer> in_gpu_range = { 0 };
+	in_gpu_range.offset = begin * sizeof(SortElement);
+	in_gpu_range.size = (end - begin) * sizeof(SortElement);
+	in_gpu_range.buffer = in_gpu;
+
+	auto sorter = core::make_smart_refctd_ptr<RadixSortClass>(driver, WG_SIZE);
+	auto scanner = core::make_smart_refctd_ptr<ScanClass>(driver, ScanClass::Operator::ADD, WG_SIZE);
+
+	auto ds_histogram = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		sorter->getDefaultHistogramDescriptorSetLayout()));
+	auto histogram_pipeline = sorter->getDefaultHistogramPipeline();
+
+	auto ds_upsweep = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		scanner->getDefaultDescriptorSetLayout()));
+	auto upsweep_pipeline = scanner->getDefaultUpsweepPipeline();
+
+	auto ds_downsweep = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		scanner->getDefaultDescriptorSetLayout()));
+	auto downsweep_pipeline = scanner->getDefaultDownsweepPipeline();
+
+	auto ds_scatter = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		sorter->getDefaultScatterDescriptorSetLayout()));
+	auto scatter_pipeline = sorter->getDefaultScatterPipeline();
+
+	{
+		driver->beginScene(true);
+
+		std::cout << "GPU sort begin" << std::endl;
+
+		auto start = std::chrono::high_resolution_clock::now();
+
+		RadixSort(driver, in_gpu_range, ds_histogram.get(), histogram_pipeline, ds_upsweep.get(), upsweep_pipeline,
+			ds_downsweep.get(), downsweep_pipeline, ds_scatter.get(), scatter_pipeline);
+
+		auto stop = std::chrono::high_resolution_clock::now();
+
+		std::cout << "GPU sort end\nTime taken: " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << " microseconds" << std::endl;
+
+		driver->endScene();
+	}
+
+	{
+		std::cout << "CPU sort begin" << std::endl;
+
+		auto start = std::chrono::high_resolution_clock::now();
+		std::stable_sort(in + begin, in + end, [](const SortElement& a, const SortElement& b) { return a.key < b.key; });
```

Comment:
use `core::radix_sort` for timeing comparisons

Reply:
you need an apples to apples (O(n) to O(n)) comparison not apples to oranges (O(n) to (n log n))

Reply:
I don't think `core::radix_sort` works perfectly yet

Reply:
definitely doesn't if the bugs you've reported are to be believed. I'll test and fix

Reply:
fixed the bugs

Reply:
works like a charm!